### PR TITLE
#4 Module 5 - Bloc pattern and Unit test

### PR DESCRIPTION
**Pull request by agustingodoyc**
The main objective of this exercise is to introduce the API Rest calls from the Flutter side, using HTTP Protocol.
https://drive.google.com/file/d/1UEv9CWHbBSMXscUNK32td1goK5HMXqO1/view?usp=sharing

### CODE REVIEW

File: lib/main.dart [hunk 1]
```diff
@@ -1,15 +1,22 @@
...
 import 'src/config/router/app_router.dart';
+import 'src/config/themes/app_theme.dart';
 
-void main() => runApp(const MyApp());
+void main() => runApp(
+      const MyApp(),
+    );
 
-class MyApp extends StatelessWidget {
+class MyApp extends StatefulWidget {
```
Comment (matiasilveiro-gl) [inline lib/main.dart:11]: This could be `Stateless`

File: lib/src/core/utils/constants/api_strings.dart [hunk 2]
```diff
@@ -0,0 +1,17 @@
+class ApiStrings {
+  static const String homeRoute = '/';
+  static const String movieRoute = '/movie_details';
+  static const String pageRoute = '/movies_page';
```
Comment (matiasilveiro-gl) [inline lib/src/core/utils/constants/api_strings.dart]: These are not related to the `Api`, so they'd be better in a `NavigationRoutes` class or so

File: lib/src/data/repositories/repository.dart [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+class Repository implements IGenresRepository, IMoviesRepository {
+  final ApiService _apiService;
+
+  Repository({ApiService? apiService})
+      : _apiService = apiService ?? ApiService();
+
+  @override
+  Future<DataState<List<GenreModel>>> fetchGenres() async {
+    DataState<dynamic> result = await _apiService.fetchGenres();
+    if (result.state == StateEnum.success) {
```
Comment (matiasilveiro-gl) [inline lib/src/data/repositories/repository.dart]: This could be `if (result.state is DataSuccess)` instead!
The same for the rest

File: lib/src/presentation/widgets/movie_details/movie_header.dart [hunk 4]
```diff
@@ -22,8 +23,19 @@ class MovieHeader extends StatelessWidget {
...
+                context,
+                url,
+              ) =>
+                  Image.asset('assets/images/placeholder.jpg'),
+              errorWidget: (
+                context,
+                url,
+                error,
+              ) =>
+                  Image.asset('assets/images/placeholder.jpg'),
```
Comment (MarcosCGlobant) [inline lib/src/presentation/widgets/movie_details/movie_header.dart]: this assets should be saved on a constant

File: lib/src/presentation/widgets/movie_details/movie_title.dart [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+                fontWeight: FontWeight.w500,
+              ),
+            ),
+            TextSpan(
+              text: ' ($year)',
+              style: TextStyle(
+                color: textColor.withOpacity(lessOpacity),
+                fontSize: AppDimens.headingFontSize,
+                fontWeight: FontWeight.w500,
+              ),
```
Comment (MarcosCGlobant) [inline lib/src/presentation/widgets/movie_details/movie_title.dart]: this could be abstracted into a new widget

File: lib/src/presentation/views/home.dart [hunk 6]
```diff
@@ -0,0 +1,32 @@
+import 'package:flutter/material.dart';
+
+import '../../core/utils/enums/page_enum.dart';
+import '../widgets/general/app_drawer/app_drawer.dart';
+import '../widgets/home/page_preview.dart';
+
+class Home extends StatefulWidget {
```
Comment (matiasilveiro-gl) [inline lib/src/presentation/views/home.dart]: This could be `Stateless`

File: lib/src/presentation/views/movies_page.dart [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+  @override
+  State<StatefulWidget> createState() => _MoviesPageState();
+}
+
+class _MoviesPageState extends State<MoviesPage> {
+  @override
+  Widget build(BuildContext context) {
+    final PageEnum pageEnum =
+        ModalRoute.of(context)!.settings.arguments as PageEnum;
+    final MoviesBloc moviesBloc = MoviesBloc();
```
Comment (matiasilveiro-gl) [inline lib/src/presentation/views/movies_page.dart]: Initialize the bloc in the `initState`! You're creating a new bloc each time the widget is rebuilt

File: lib/src/presentation/widgets/home/page_preview.dart [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+  });
+
+  @override
+  State<StatefulWidget> createState() => _PagePreviewState();
+}
+
+class _PagePreviewState extends State<PagePreview> {
+  @override
+  Widget build(BuildContext context) {
+    final MoviesBloc moviesBloc = MoviesBloc();
```
Comment (matiasilveiro-gl) [inline lib/src/presentation/widgets/home/page_preview.dart]: Same as before

File: lib/src/presentation/widgets/movie_details/details_text/details_box.dart [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+            Text(
+              releaseDate,
+              style:
+                  TextStyle(color: Theme.of(context).colorScheme.onPrimary),
+            ),
+            Text(
+              genres,
+              style:
+                  TextStyle(color: Theme.of(context).colorScheme.onPrimary),
+            ),
```
Comment (matiasilveiro-gl) [inline lib/src/presentation/widgets/movie_details/details_text/details_box.dart]: Trailing commas

File: lib/src/presentation/widgets/movie_details/movie_title.dart [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+    const lessOpacity = 0.6;
+    return Padding(
+      padding: const EdgeInsets.only(
+        top: padding,
+        left: padding,
+        right: padding,
+      ),
+      child: RichText(
+        textAlign: TextAlign.center,
+
```
Comment (matiasilveiro-gl) [inline lib/src/presentation/widgets/movie_details/movie_title.dart]: Remove this space

File: test/widget_testing/home_test/preview_header_test/page_navigator_test.dart [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+import 'package:flutter_test/flutter_test.dart';
+import 'package:the_movie_db_flutter/src/core/utils/enums/page_enum.dart';
+import 'package:the_movie_db_flutter/src/presentation/widgets/home/preview_header/page_navigator.dart';
+
+void main() {
+  testWidgets(
+    'PageNavigator Test',
+        (WidgetTester tester) async {
+      const mockPage = PageEnum.popular;
+      await tester.pumpWidget(
```
Comment (matiasilveiro-gl) [inline test/widget_testing/home_test/preview_header_test/page_navigator_test.dart:27]: You could also test the navigation when tapped!

File: lib/src/core/utils/resources/network_image_util.dart [hunk 12]
```diff
@@ -12,11 +12,9 @@ class NetworkImageUtil {
       );
       return newColorScheme;
     } catch (e) {
-      return ColorScheme.fromSwatch();
+      ImageProvider imageProvider =
+          const AssetImage('assets/images/placeholder.jpg');
```
Comment (Kevin-Guillen) [inline lib/src/core/utils/resources/network_image_util.dart]: constant

File: lib/src/presentation/widgets/movie_details/details_vote_average.dart [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+    super.key,
+    required this.voteAverage,
+  });
+
+  final num voteAverage;
+
+  @override
+  Widget build(BuildContext context) {
+    const int maxLines = 2;
+    const double textWidth = 55.0;
```
Comment (Kevin-Guillen) [inline lib/src/presentation/widgets/movie_details/details_vote_average.dart]: move this consts outside the build

File: lib/main.dart [hunk 14]
```diff
@@ -1,21 +1,47 @@
 import 'package:flutter/material.dart';
-import 'package:the_movie_db_flutter/src/config/themes/app_theme.dart';
 
 import 'src/config/router/app_router.dart';
+import 'src/config/themes/app_theme.dart';
+import 'src/core/utils/resources/color_scheme_util.dart';
```
Comment (FranciscoGaraventa) [inline lib/main.dart:5]: relative imports

File: lib/src/core/parameter/parameter_object.dart
Comment (FranciscoGaraventa) [inline lib/src/core/parameter/parameter_object.dart:1]: If you don't use this file, you can remove it

File: lib/src/core/utils/app_constants/app_dimens/font_sizes.dart
Comment (FranciscoGaraventa) [inline lib/src/core/utils/app_constants/app_dimens/font_sizes.dart:1]: Instead to made this as a class you can replace it with
```
abstract class FontSizes {
   ...
}
```
With this change you don't need to add a const constructor

File: lib/src/core/utils/app_constants/app_routes/api_routes.dart [hunk 15]
```diff
@@ -0,0 +1,17 @@
+class ApiRoutes {
+  const ApiRoutes();
```
Comment (FranciscoGaraventa) [inline lib/src/core/utils/app_constants/app_routes/api_routes.dart]: Same as above

File: lib/src/core/utils/app_constants/app_routes/app_routes.dart [hunk 16]
```diff
@@ -0,0 +1,7 @@
+import 'api_routes.dart';
+import 'navigation_routes.dart';
+
+class AppRoutes {
+  static const ApiRoutes api = ApiRoutes();
+  static const NavigationRoutes navigation = NavigationRoutes();
+}
```
Comment (FranciscoGaraventa) [inline lib/src/core/utils/app_constants/app_routes/app_routes.dart]: EOF

File: lib/src/core/utils/app_constants/app_strings/error_message.dart [hunk 17]
```diff
@@ -0,0 +1,7 @@
+class ErrorMessage {
+  const ErrorMessage();
+
+  static const String error = 'Error';
+  static const String empty = 'Empty data';
+  static const String other = 'other';
+}
```
Comment (FranciscoGaraventa) [inline lib/src/core/utils/app_constants/app_strings/error_message.dart]: Instead to be a class you can use a enum class
```
enum class ErrorMessages {
    error('Error),
    empty('Empty data'),
    other('Other');

    const ErrorMessages(this.msg)
    
    final String msg;
}
```

File: lib/src/presentation/views/movies_page.dart [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+import '../../core/utils/enums/page_enum.dart';
+import '../blocs/movies_bloc.dart';
+import '../widgets/general/app_drawer/app_drawer.dart';
+import '../widgets/general/page_title.dart';
+import '../widgets/movie_page/movies_column.dart';
+
+class MoviesPage extends StatefulWidget {
+  final PageEnum pageEnum;
+
+  const MoviesPage({super.key, required this.pageEnum});
```
Comment (FranciscoGaraventa) [inline lib/src/presentation/views/movies_page.dart]: trailing comma

File: lib/src/presentation/widgets/general/app_drawer/app_drawer.dart [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+            children: [
+              const Padding(
+                padding: EdgeInsets.symmetric(
+                  vertical: AppPaddings.textsLarge,
+                  horizontal: AppPaddings.textsLarge,
+                ),
+                child: DrawerTitle(title: AppTitles.movies),
+              ),
+              ...PageEnum.values.map(
+                (e) => Padding(
```
Comment (FranciscoGaraventa) [inline lib/src/presentation/widgets/general/app_drawer/app_drawer.dart]: what is 'e'?

File: lib/src/presentation/widgets/general/poster_image.dart [hunk 20]
```diff
@@ -0,0 +1,30 @@
...
+      imageUrl: imageUrl,
+      placeholder: (
+        context,
+        url,
+      ) =>
+          Image.asset('assets/images/placeholder.jpg'),
+      errorWidget: (
+        context,
+        url,
+        error,
```
Comment (FranciscoGaraventa) [inline lib/src/presentation/widgets/general/poster_image.dart]: Add the type of the parameters

File: lib/src/presentation/widgets/home/movies_row/movies_row.dart [hunk 21]
```diff
@@ -0,0 +1,83 @@
...
+    super.initState();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return StreamBuilder(
+      stream: widget.moviesBloc.moviesStream,
+      builder: (
+        context,
+        snapshot,
```
Comment (FranciscoGaraventa) [inline lib/src/presentation/widgets/home/movies_row/movies_row.dart]: Add parameter types

**Review by MarcosCGlobant — APPROVED** [review]

**Review by Kevin-Guillen — APPROVED** [review]

**Review by matiasilveiro-gl — APPROVED** [review]

**Review by FranciscoGaraventa — APPROVED** [review]
